### DISCUSSION TITLE
Many Unknown values for Q-Cells/Solax with serial H34C12Ixxxxxx

### DISCUSSION THREAD

**Discussion by plu** [wills106/homeassistant-solax-modbus#592] [Q&A]
Some values are working, for example the serial number, and some of the setters are also working. But most values just show up as `Unknown`. I've enabled the debug log and attached the log file here: 

[home-assistant_solax_modbus_2023-11-03T07-54-57.286Z.log](https://github.com/wills106/homeassistant-solax-modbus/files/13247887/home-assistant_solax_modbus_2023-11-03T07-54-57.286Z.log)

I'm connecting via ModBus TCP. The COM-Port is exposed via TCP using some ESP8266 using [Tasmota's ModBus TCP Bridge](https://tasmota.github.io/docs/Modbus-Bridge/#additional-commands-for-use_modbus_bridge_tcp).

The inverter reports following version:

* DSP-Version: 2.07
* ARM-Version: 2.03

HA version:

* Home Assistant 2023.10.2
* Frontend 20231005.0 - latest
* homeassistant-solax-modbus - just checked out the git repository at 85a611724a78e63990ece6a60bbfbed71899c6cc

I'm not even sure where to start debugging this further without any help, would be great if somebody could give me some pointers. If there's more information that I could or should provide, please let me know.

**Comment by wills106** [comment]
Do they look like the bottom two?

![image](https://github.com/wills106/homeassistant-solax-modbus/assets/18155231/b3d9ab20-0aba-4e4c-b40f-6945566ac138)
Then when you click on them...
![image](https://github.com/wills106/homeassistant-solax-modbus/assets/18155231/b2024610-c1fa-442e-8ae9-22afd21b7b20)
If so look at point 1 https://github.com/wills106/homeassistant-solax-modbus#faq

**Reply by plu** [reply]
Nope, not hidden ones, they really do show `Unknown`.

![Screenshot 2023-11-03 at 10 07 59](https://github.com/wills106/homeassistant-solax-modbus/assets/31597/b5fc37d3-370b-4ab0-b339-18f6d6714a52)

**Reply by plu** [reply]
I cannot shake the feeling that it's related to my Tasmota ModBus TCP Bridge. But I cannot see any error message there, neither in HomeAssistant. On the other hand I have a similar setup running for years to read values from my Waterkotte heat pump, without any issues.

**Reply by plu** [reply]
I've also started looking for random failures, so I came up with this script:

```python
#!/usr/bin/env python
from pymodbus.client import ModbusTcpClient
from pymodbus.payload import BinaryPayloadDecoder, Endian

c = ModbusTcpClient(host="192.168.6.82", port=4444, timeout=5)
c.connect()
inverter_data = c.read_holding_registers(0x0, 7, slave=1)
decoder = BinaryPayloadDecoder.fromRegisters(inverter_data.registers, byteorder=Endian.BIG)
serial = decoder.decode_string(14).decode("ascii")
print(serial)
```

And I ran that for a while in a loop:

```shell
while true; do ./test.py; done
```

And it never fails, it always prints the serial of my inverter successfully.

**Reply by wills106** [reply]
Looking at your log the Integration is going into sleep mode.
So either you Inverter is sleeping due to lack of PV / Battery flat or your not getting a proper response from the Inverter through the bridge. I don't have any experience of using that connection method.

**Reply by plu** [reply]
When the plugin connects to the TCP bridge, how many requests is it usually sending to the bus?

On the Tasmota device I always see two responses (`ModbusReceived`) after the plugin (`192.168.10.10`) connected:

```
10:31:31.578 TCP: MBS: MBRTCP Got connection from 192.168.10.10
10:31:31.691 MQT: tele/tasmota/qcells/RESULT = {"ModbusReceived":{"DeviceAddress":1,"FunctionCode":3,"StartAddress":0,"Length":19,"Count":7,"Values":[18483,..............................................]}}
10:31:31.816 MQT: tele/tasmota/qcells/RESULT = {"ModbusReceived":{"DeviceAddress":1,"FunctionCode":3,"StartAddress":125,"Length":131,"Count":63,"Values":[26,0,0,0,0,6,........................................,12000,1]}}
10:32:29.496 MQT: tele/tasmota/qcells/STATE = {"Time":"2023-11-03T10:32:29","Uptime":"0T00:53:11","UptimeSec":3191,"Heap":15,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Wifi":{"AP":1,"SSId":"...........","BSSId":"02:.........","Channel":1,"Mode":"11n","RSSI":60,"Signal":-70,"LinkCount":1,"Downtime":"0T00:00:06"}}
```

Unfortunately Tasmota can only log the responses, not the requests. But if you say there's more than two requests per cycle, then I'll start digging into that direction.

**Reply by wills106** [reply]
On the Gen4 SolaX there are 6 Blocks it reads you can see them in your log:

```
2023-11-03 07:54:02.160 INFO (MainThread) [custom_components.solax_modbus.sensor] SolaX returning holding block: 0x0 0x7 [0]
2023-11-03 07:54:02.160 INFO (MainThread) [custom_components.solax_modbus.sensor] SolaX returning holding block: 0x7d 0xbc [125, 126, 127, 128, 129, 130, 131, 132, 133, 139, 140, 141, 142, 143, 144, 145, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 161, 162, 163, 164, 165, 166, 167, 170, 178, 181, 182, 183, 186, 187]
2023-11-03 07:54:02.160 INFO (MainThread) [custom_components.solax_modbus.sensor] SolaX returning holding block: 0x102 0x162 [258, 259, 260, 261, 262, 263, 264, 265, 266, 270, 273, 275, 277, 278, 279, 280, 283, 284, 285, 286, 287, 288, 289, 290, 291, 292, 293, 294, 295, 296, 297, 303, 304, 305, 306, 335, 336, 337, 338, 339, 340, 341, 342, 343, 344, 352, 353]
2023-11-03 07:54:02.160 INFO (MainThread) [custom_components.solax_modbus.sensor] SolaX returning input block: 0x0 0x55 [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 19, 20, 21, 22, 23, 24, 26, 28, 29, 32, 33, 35, 36, 37, 38, 70, 72, 74, 79, 80, 82, 84]
2023-11-03 07:54:02.160 INFO (MainThread) [custom_components.solax_modbus.sensor] SolaX returning input block: 0x66 0xcb [102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 132, 134, 136, 138, 142, 144, 145, 146, 148, 150, 152, 154, 168, 170, 172, 174, 176, 178, 180, 182, 184, 185, 200, 201, 202]
2023-11-03 07:54:02.160 INFO (MainThread) [custom_components.solax_modbus.sensor] SolaX returning input block: 0xcb 0x116 [203, 204, 256, 257, 258, 260, 262, 264, 266, 268, 270, 272, 276]
```

It reads them at the polling rate you set.
On the bridge do you have the Baud set at 19200?

**Reply by plu** [reply]
Hmm, ok - thank you!

> On the bridge do you have the Baud set at 19200?

Correct, bridge has 19200 set, and on the inverter it's the same. Parity is 8N1.

**Reply by plu** [reply]
I'm still confused why I only see two `ModbusReceived` every time the plugin connects to the bridge. Maybe my next step should be to add some debug logging to Tasmota to also log what values it's actually sending onto the bus.

**Reply by plu** [reply]
Thank you for all your help so far, and for developing and maintaining this plugin!

**Reply by plu** [reply]
```python
from pymodbus import pymodbus_apply_logging_config
pymodbus_apply_logging_config("DEBUG")
```

I've added this now to `custom_components/solax_modbus/__init__.py`, and the logs produced by this are looking all just fine to me: [pymodbus.log](https://github.com/wills106/homeassistant-solax-modbus/files/13249563/pymodbus.log)

If you spot something unusual, please let me know. I've tried to mask some of the response by replacing values with `.............`.

The way I understand your plugin so far is that the model is being detected by the serial number, right? I'm wondering if that's what might be going wrong in my case, that it picks the wrong model and then uses the wrong modbus addresses?

**Reply by plu** [reply]
Looking more closely again at the logs I've noticed this:

```
[pymodbus.logging] Not a valid slave id - 0, ignoring!!
```

Not sure how much that might play a role...

**Reply by plu** [reply]
Got some Waveshare adapter yesterday, and now everything works \o/.

![Screenshot 2023-11-05 at 08 58 26](https://github.com/wills106/homeassistant-solax-modbus/assets/31597/f15259dc-7afd-41f6-aa05-4ef7100a6b20)

If I'm really bored, I might be poking around again with Tasmota, as I'd prefer that solution, but for now we can say "case closed". Thanks again for your help and for making this plugin!
